### ISSUE TITLE
gem installation is broken with an SSL error

### ISSUE THREAD

**Issue by david-rui** [oneclick/rubyinstaller#338]
I installl ruby and try install gem of compass but SSL is not working.

C:\Users\david.ruiz>gem install compass
ERROR:  Could not find a valid gem 'compass' (>= 0), here is why:
          Unable to download data from https://rubygems.org/ - SSL_connect returned=1 errno=0 state=SSLv3 read server certificate B: certificate verify failed (https://api.rubygems.org/specs.4.8.gz)

**Comment by Azolo** [comment]
This is actually a new issue that is similar to an old problem. Look at [this comment on a rubygems issue](https://github.com/rubygems/rubygems/issues/1736#issuecomment-252074698) for more information.

The easy, not-recommended, and insecure fix is:

```
gem install rubygems-update --source "http://rubygems.org"
update_rubygems
gem uninstall rubygems-update -x
```

The secure and recommended fix for this is actually using @luislavena's [detailed post on updating](https://gist.github.com/luislavena/f064211759ee0f806c88#installing-using-update-packages-new) except getting the most recent rubygems release manually at https://rubygems.org/downloads/rubygems-update-2.6.7.gem.

**Comment by robgha01** [comment]
Im having the exact same problem reinstalled ruby 2.2.5 from 2.3 still same...

**Comment by luislavena** [comment]
Install RubyGems manually:

- Use the browser to download `.gem` file from RubyGems installation page: https://rubygems.org/pages/download
- Remember where file was downloaded (Say `C:\Users\Luis\Downloads`)
- Open a new *Start Command Prompt with Ruby*
- Install downloaded gem using the following command: `gem install --local rubygems-update-2.6.7.gem`
- Once it completes, run `update_rubygems`
- Check installation succeeded by checking the output of `gem --version`, It must report the same version that was downloaded above.

All this is described in the links from [previous comment](https://github.com/oneclick/rubyinstaller/issues/338#issuecomment-252077104).

RubyGems performed *another* upgrade of their infrastructure and affected users that didn't upgrade RubyGems in time.

(This is also described in the gist linked in previous comment from a previous upgrade)

**Comment by luislavena** [comment]
This has been covered by official RubyGems documentation:

http://guides.rubygems.org/ssl-certificate-update/

See *INSTALLING USING UPDATE PACKAGES (NEW)* section.

**Comment by SlaWitDev** [comment]
Hi, Yesterday I had to reinstall ruby (2.3.1-p112-x64) on Windows 10 x64 (1607/anniversary build).

gem install anything or gem update --system doesn't work cause "SSL_connect returned=1 errno=0 state=SSLv3 read server certificate B: certificate verify failed"

After few hours I found solution, secure and easy and I think it's good to put it into rubyinstaller.

Trouble is in openssl / windows 10 / certyficates. PEM files in rubygems/ssl_certs are ok. Instruction from http://guides.rubygems.org/ssl-certificate-update/ works, but only with manual (--local) gem instalation / or update but it doesn't fix the problem, or another words "Step 3: Copy new trust certificate" doesn't enought (in rubyinstaller-2.3.1-x64.exe you have actual certificates just after install - don't need this step).

I found is a trouble with openssl (in this case OpenSSL 1.0.1l 15 Jan 2015 [ you can check this by: ruby -ropenssl -e 'puts OpenSSL::OPENSSL_VERSION' ])
I don't known why but this openssl (or may be on Windows 10 openssl) don't known any of trusted CA certyficates.

Siple solution is set ENV VAR:
set SSL_CERT_DIR=**c:\ruby**\lib\ruby\2.3.0\rubygems\ssl_certs\
or for single one:
set SSL_CERT_FILE=**c:\ruby**\lib\ruby\2.3.0\rubygems\ssl_certs\**AddTrustExternalCARoot-2048.pem**

**c:\ruby** is my instalation dir and **AddTrustExternalCARoot-2048.pem** is installed and current CACertFile it means you in rubyinstaller-2.3.1-x64.exe you don't nedd obtain the new trust certificate cause you have it, you must tune part of value **c:\ruby** for your instalation path.

Probably is the solution for ruby / gem on Windows 10 for others versions to and it can be set by ruby installer.

Best regards
SlaWitDev

**Comment by Azolo** [comment]
@SlaWitDev There have been multiple conversations about shipping the root trust certificates for Ruby, however the decision that was made is that we shouldn't handle CA Root certificates.

I wish I could link one of those conversations but I don't have time to find them at the moment.

**Comment by SlaWitDev** [comment]
@Azolo. OK set ENV VAR in my opinion isn't shipping certificates, but OK.

In this case short way to fix SSL errors for Windows 10, we need manual configuration, cause:
openssl for ruby (shipped by rubyinstaller-2.3.1-x64)  looks for certificates in:
_- C:/Users/Justin/Projects/knap-build/var/knapsack/software/x64-windows/openssl/1.0.1l/ssl/private
- C:/Users/Justin/Projects/knap-build/var/knapsack/software/x64-windows/openssl/1.0.1l/ssl
- C:/Users/Justin/Projects/knap-build/var/knapsack/software/x64-windows/openssl/1.0.1l/ssl/certs
- C:/Users/Justin/Projects/knap-build/var/knapsack/software/x64-windows/openssl/1.0.1l/ssl/cert.pem
**- %SSL_CERT_DIR%**
**- %SSL_CERT_FILE%**
We can set locally in system only SSL_CERT_DIR or SSL_CERT_FILE. And for fresh install or reinstall ruby need it, as I wrote before.

**Comment by felixfbecker** [comment]
I understand that some manual update steps may be required when I have rubygems already installed because of the change in the infrastructure. But why do I need to "update" anything in a fresh Ruby installation?

**Comment by luislavena** [comment]
> But why do I need to "update" anything in a fresh Ruby installation?

RubyInstaller bundles an specific version of RubyGems, which also bundles the SSL certificates required to connect to rubygems.org.

Is quite common to update bundled RubyGems from time to time (there is no perfect software) but the installers are paired/locked to the version of Ruby: there are no new installers when RubyGems changes, only when Ruby does.

The installers were released some point before the new version of RubyGems introduced the new certificates, thus not allowing you to connect to rubygems.org with the bundled version.

This becomes a chicken an egg situation: you cannot update RubyGems using `gem update --system` because it doesn't have the cert to connect to rubygems.org, so the manual installation of the update (following the guides) is the only approach.

If you check the number of releases of RubyGems, you will notice that they happen more often than Ruby itself (every couple of months vs once or twice a year for Ruby). Each installer release takes considerable amount of time from maintainers to test.

Hope that answer your question.

**Comment by felixfbecker** [comment]
Thanks for the detailed answer. I really feel in this case they should have coordinated the changes/updates in a way that doesn't break fresh installations.

**Comment by luislavena** [comment]
@felixfbecker you're preaching to the choir here, but even so, coordinate such a massive upgrade across many distributions of Ruby (because RubyInstaller is not the only one) is way more complicated than you can think of.

And even in such case, users do not update their Ruby, they will reach out here, report they have a problem and **then** update the installation of Ruby, so we are again on a similar scenario.

**Comment by deveth0** [comment]
Please also see this Workaround for the issue: 
https://github.com/rubygems/rubygems/issues/1736#issuecomment-261862935

tl;dr: replace https source with http, update, switch back.

**Comment by anderson109119** [comment]
> Install RubyGems manually:
> 
> * Use the browser to download `.gem` file from RubyGems installation page: https://rubygems.org/pages/download
> * Remember where file was downloaded (Say `C:\Users\Luis\Downloads`)
> * Open a new _Start Command Prompt with Ruby_
> * Install downloaded gem using the following command: `gem install --local rubygems-update-2.6.7.gem`
> * Once it completes, run `update_rubygems`
> * Check installation succeeded by checking the output of `gem --version`, It must report the same version that was downloaded above.
> 
> All this is described in the links from [previous comment](https://github.com/oneclick/rubyinstaller/issues/338#issuecomment-252077104).
> 
> RubyGems performed _another_ upgrade of their infrastructure and affected users that didn't upgrade RubyGems in time.
> 
> (This is also described in the gist linked in previous comment from a previous upgrade)

Good, good,man, This command local help me. Thank you.
